Add distributed mesh support and fake neighbor handling in BreakMeshByBlock

ChengHauYang opened the pull request:
closes #12033
closes #21161

Previous PR #31636 had many repeated messages from the moosebuild bot due to a temporary CIVET hiccup, making the thread difficult to review. As suggested, I am opening a new clean PR here and closing the previous one.

## Reason

Currently, **BreakMeshByBlock (bmbb)** uses a hacky approach to prepare the mesh.  
When nodes are duplicated by bmbb, their former neighbors are no longer neighbors (since neighborhood is based on shared nodes). As a result:

- We cannot call `prepare_for_use()` at the end of bmbb, because it would erase the discontinuous neighbor information.
- To work around this, the framework has special hacks for bmbb (e.g., skipping mesh validity checks in debug mode when bmbb is present).

Another limitation is that bmbb cannot run with **distributed meshes**. This PR addresses that.

---

## Design

1. **Preserving discontinuous neighbors**  
   - Since `prepare_for_use()` wipes out discontinuous neighbor data, we introduce a new data structure to store this information (referred to as *fake neighbors*).  
   - These fake neighbors remain accessible to interface kernels that require them (e.g., CZM).

2. **Support for distributed meshes**  
   Following Roy’s suggestions:  
   - **Unique node IDs**: In replicated meshes, new nodes were created with  
     ```cpp
     Node::build(*current_node, mesh->n_nodes()).release();
     ```  
     This fails in distributed meshes.  
     -> Solution: use communication between processors and assign deterministic IDs based on `max_node_id` strides.  
   - **Ghost node handling**:  
     - *Phase 0*: Each rank computes locally connected blocks for its owned nodes and sends this info to the node owner.  
     - *Phase 1*: Node owners aggregate connected block info and broadcast the global set to all ranks.  
   - **Boundary ID assignment**: Updated in line with Roy’s fix ([commit link](https://github.com/roystgnr/moose/commit/2583950d46aa79a928903f392df908f4678f18d3)).

---

## Impact

- bmbb now supports **distributed meshes**.  
- `prepare_for_use()` can be safely called at the end of bmbb.  
- Removes the need for special hacks in the MOOSE framework for bmbb.

ChengHauYang committed:
Fix BreakMeshByBlockGenerator for distributed meshes and add FakeNeighbor RM

1. Distributed mesh fixes
   - Node ID assignment: using `mesh->n_nodes()` fails for distributed meshes (IDs are sparse). Switch to `max_node_id` with a stride.
   - Ghost node handling:
     - *Phase 0*: Each rank computes locally connected blocks for its owned nodes and sends them to the node owner.
     - *Phase 1*: Node owners aggregate connected block info and broadcast the global set to all ranks.
   - Boundary ID assignment aligned with Roy’s fix ([commit link](https://github.com/roystgnr/moose/commit/2583950d46aa79a928903f392df908f4678f18d3)).

2. New FakeNeighbor datastructure
   - Introduce a RelationshipManager for ghosted fake neighbors (CZM discontinuous neighbors, not topological).
   - Robust initialization:
     - Override `set_mesh` to ensure `_functor` is created JIT on first use.
     - Rebuild `_functor` in `internalInitWithMesh` if the neighbor map was populated after an early empty init.
     - Handles mesh cloning (`SetupMeshAction`) cases where `set_mesh` is called before initialization.

3. Simplifications
   - Since we now call `prepare_for_use()` at the end of BMBb, remove BreakMeshByBlockGenerator-specific hacks.
   - Simplify node ID handling.
   - Drop `MeshGeneratorSystem::_has_bmbb` flag and `hasBreakMeshByBlockGenerator()` API.

# Conflicts:
#	framework/src/base/MeshGeneratorSystem.C

# Conflicts:
#	framework/include/mesh/MooseMesh.h

ChengHauYang committed:
Always create map entries for nodes in _nodeid_to_connected_blocks

Fixes crashes from map::at when nodes had no valid blocks. Each node now
gets an entry (even empty), keeping the map consistent with node_to_elem_map.

ChengHauYang committed:
revert some changes so that the we do not create unwanted dulplicated nodes

ChengHauYang committed:
This commit fixes a segmentation fault when using block_pairs in distributed mode.

The crash was caused by an incorrect decision on node duplication, which stemmed from using incomplete, processor-local data (`node_to_elem_map`). The fix ensures the decision is based on globally synchronized data (`_nodeid_to_connected_blocks`), guaranteeing a consistent and correct outcome across all processors.

ChengHauYang committed:
If we set block_pairs in bmbb, the node should not duplicate when the node is in a junction with more than two blocks that share the node. It only happens that the node is in the interface of two blocks and has the block IDs equal to what the user set.

ChengHauYang committed:
closes #12033 and #21161 (1) remove REPLICATED mode restriction for bmbb tests (2) regold for the bmbb tests

ChengHauYang committed:
[WIP] change the neighbor ptr implementation to one line

ChengHauYang committed:
[WIP] API for fake neighbor

hugary1995 committed:
slightly changed the API in MooseMesh related to disconnected neighbors

ChengHauYang committed:
Implement a parallel-safe disconnected neighbor system

Refactors the disconnected neighbor mechanism to be robust and safe under parallel (MPI) execution.

Previously, the implementation stored raw `Elem*` pointers in a map, which was fundamentally unsafe in a distributed environment. Pointers are only valid on the local rank and become meaningless across MPI processes, leading to segmentation faults in multi-process runs.

This commit resolves the issue by transitioning to an ID-based design:

- Core map (`_disconnected_neighbors_by_id`) now stores `dof_id_type` pairs instead of raw `ConstBndElement` pointers, ensuring portability and serializability across ranks.
- The `disconnectedNeighbor` accessor now returns a   `std::optional<ConstBndElement>`, replacing unsafe raw pointers with a safer C++ interface that naturally handles missing neighbors.

**Key Learning:** For parallel-safe design, always **store globally unique
IDs** and resolve them into rank-local pointers, which avoids pointer validity issues and ensures correctness across distributed
executions.

ChengHauYang committed:
clean the code

hugary1995 committed:
change data structure

ChengHauYang committed:
refactor(mesh): disconnected neighbor handling

- Replaced raw Elem* usage with ID-based ElemSide pairs for safer parallel handling
- Added cache (_cached_disconnected_neighbors) to speed up repeated lookups
- Modified neighbor queries across loops (NonlinearThread, ComputeUserObjectsThread, ComputeMaterialsObjectThread, FEProblemBase, etc.) to use disconnectedNeighborPtr

ChengHauYang committed:
remove the place the do not need disconnectedNeighbor

ChengHauYang committed:
change libmesh side to reach disconnected neighbor through neighbor_ptr

ChengHauYang committed:
works for opt mode (not for dbg mode)

ChengHauYang committed:
update datastructure in libmesh

ChengHauYang committed:
support distributed meshes

ChengHauYang committed:
update libmesh to use disconnected boundary (libmesh commit: bee28f5)

ChengHauYang committed:
czm tests do not limit to replicated mesh only

ChengHauYang committed:
checkout libmesh

# Conflicts:
#	libmesh

ChengHauYang committed:
fix the unit tests

ChengHauYang committed:
checkout libmesh to devel

ChengHauYang committed:
slighly clean the unused functions

ChengHauYang committed:
add back missing APIs after rebasing

ChengHauYang committed:
regold one file

GiudGiud commented on framework/src/constraints/EqualValueBoundaryConstraint.C:
```diff
@@ -88,9 +94,59 @@ EqualValueBoundaryConstraint::updateConstrainedNodes()
   _primary_node_vector.clear();
   _connected_nodes.clear();
 
+  // user provide nothing
```
```suggestion
  // user provided nothing
```

GiudGiud commented on framework/src/constraints/EqualValueBoundaryConstraint.C:
```diff
@@ -58,6 +58,11 @@ EqualValueBoundaryConstraint::validParams()
+  params.addParam<Point>("primary_node_coord",
+                         Point(std::numeric_limits<unsigned int>::max(),
+                               std::numeric_limits<unsigned int>::max(),
+                               std::numeric_limits<unsigned int>::max()),
```
```suggestion
                         Point(std::numeric_limits<Real>::max(),
                               std::numeric_limits<Real>::max(),
                               std::numeric_limits<Real>::max()),
```

ChengHauYang replied:
Fixed!

GiudGiud commented on framework/src/constraints/EqualValueBoundaryConstraint.C:
```diff
@@ -88,9 +94,59 @@ EqualValueBoundaryConstraint::updateConstrainedNodes()
     mooseError("Please specify secondary node ids or boundary id.");
-  else if ((_secondary_node_ids.size() == 0) && (_secondary_node_set_id != "NaN"))
+
+  if (_primary_node_id == std::numeric_limits<unsigned int>::max() &&
```
```suggestion
  // Find primary node id from user-specified primary node coordinates
  if (_primary_node_id == std::numeric_limits<unsigned int>::max() &&
```

ChengHauYang replied:
Fixed

GiudGiud commented on modules/solid_mechanics/test/tests/cohesive_zone_model/gold/3DC_normal_load_out.e:
are all these regolds only because of the sideset name change?
could they be avoided by using a RenameBoundaryGenerator?

GiudGiud commented on test/tests/meshgenerators/break_mesh_by_block_generator/simple_rectangle.i:
same for these 3 files I imagine

GiudGiud reviewed (commented):
early review

ChengHauYang commented on framework/src/constraints/EqualValueBoundaryConstraint.C:
```diff
@@ -88,9 +94,59 @@ EqualValueBoundaryConstraint::updateConstrainedNodes()
+
+    if (unique_count == 0)
+      mooseError("Couldn't find a node ID for the specified primary_node_coord. We go with default "
+                 "behavior of choosing primary node.");
```
Yes. Thank you for your checkings!

I just simply changed it to be

```
      mooseError("Couldn't find a node ID for the specified primary_node_coord.");

```

ChengHauYang committed:
[fix one test] The failure was caused by disjoint neighbor reconstruction being performed on the displaced mesh. Setting `use_displaced_mesh=false` keeps neighbor search on the undeformed geometry, where interface sides remain coincident, preventing Periodic boundary neighbor not found in 1D distributed meshes.

lindsayad commented on framework/include/mesh/MooseMesh.h:
```diff
@@ -1430,6 +1430,15 @@ class MooseMesh : public MooseObject, public Restartable, public PerfGraphInterf
   /// Return displace node list by side list boolean
   bool getDisplaceNodeListBySideList() { return _displace_node_list_by_side_list; }
 
+  /// Set incomplete interface pairs flag
```
pairs of what?

lindsayad replied:
You documented this in the data member, where I have a few questions/comments. Once we settle on the member doxygen, let's try to add a little more detail here for someone who encounters the method doxygen first

ChengHauYang replied:
Thanks, Alex, for your review. 

> pairs of what?

Here, the "pairs" refer to disjoint interface side pairs (the "+" and "-" sides) used to construct pseudo-neighbor relationships in libMesh. I will first update the data member doxygen to clearly define this and explain when an interface is considered incomplete (e.g., only one side provided), and add a brief, self-contained description to the getter/setter doxygen for users who encounter the API first.

We can change the doxygen later after you fully review it and have better suggestions.

lindsayad commented on framework/include/mesh/MooseMesh.h:
```diff
@@ -1430,6 +1430,15 @@ class MooseMesh : public MooseObject, public Restartable, public PerfGraphInterf
+  void setHasIncompleteInterfacePairs(bool has_incomplete)
+  {
+    _has_incomplete_interface_pairs = has_incomplete;
+  };
```
prefer that multi-line methods be outside the class definition. You can put it `inline` later in this file. Also remove semicolon at end of method definition

lindsayad commented on framework/include/mesh/MooseMesh.h:
```diff
@@ -1842,6 +1851,14 @@ class MooseMesh : public MooseObject, public Restartable, public PerfGraphInterf
   /// faces
   bool _has_lower_d;
 
+  /// Indicates whether any disconnected boundary pairs are incomplete,
```
Maybe you do mean boundary pair, but I don't think saying "disconnected" makes much sense in terms of boundary pairs. I expect just about any pair of boundaries to be disconnected

ChengHauYang replied:
Thanks, Alex. I named it "disjoint block sidesets" here ("sidesets" according to your comments below). 

It confused Roy previously (https://github.com/libMesh/libmesh/pull/4283), so we changed "disconnected" to "disjoint". Here are some quotes from him:

> I hate to bring up something huge and old on top of the issue in the new stitching support, but typing "disconnected boundary" over and over again in the midst of unrelated confusion has me wondering whether we can't come up with a better name to begin with.
> 
> Right now, if we create a slit mesh, then the boundaries on opposite sides of the slit are already topologically disconnected ... so with this PR, we can then call `add_disconnected_boundaries` to connect them, and we only refer to them as "disconnected boundaries" once they've got a connection added? It seems entirely backwards and I fear that's going to confuse users.
> 
> Got any better ideas? `disjoint_neighbor_boundary_pairs`, maybe? `disjoint_neighbor_boundaries`? Maybe the improvement of "disjoint" over "disconnected" relies on too narrow a view or too broad a background of topology here, but I think it's least a small improvement, and I definitely like getting "neighbor" in there since the principal effect of the setting is to establish neighbor links.

Sorry, I forgot to rename "disconnected" to "disjoint" in some places.

hugary1995 replied:
Removed this flag.

lindsayad commented on framework/include/mesh/MooseMesh.h:
```diff
@@ -1842,6 +1851,14 @@ class MooseMesh : public MooseObject, public Restartable, public PerfGraphInterf
   bool _has_lower_d;
 
+  /// Indicates whether any disconnected boundary pairs are incomplete,
+  /// i.e., only one side of an expected (blockA, blockB) / (blockB, blockA) pair exists.
```
In this line we're adding block language, inconsistent with the previous line. Perhaps the best language to use is "sidesets" to indicate the fact that we are pairing boundaries but that these boundaries are fundamentally linked with their element faces. Sidesets certainly fit with cohesive zone modeling using interface kernels

lindsayad replied:
Oh wait I see it below. I think you probably meant to delete this doxygen block and the `_has_incomplete_disjoint_block_sidesets` data member?

hugary1995 replied:
@lindsayad correct.

hugary1995 replied:
Removed this flag.

lindsayad commented on framework/include/meshgenerators/BreakMeshByBlockGeneratorBase.h:
```diff
@@ -56,6 +56,6 @@ class BreakMeshByBlockGeneratorBase : public MeshGenerator
 
   /// this method save the boundary name/id pair
-  void mapBoundaryIdAndBoundaryName(boundary_id_type & /*boundaryID*/,
+  void mapBoundaryIdAndBoundaryName(const boundary_id_type & /*boundaryID*/,
```
```suggestion
  void mapBoundaryIdAndBoundaryName(const boundary_id_type boundary_id,
```

ChengHauYang replied:
Fixed

lindsayad commented on framework/include/meshgenerators/BreakMeshByBlockGeneratorBase.h:
```diff
@@ -56,6 +56,6 @@ class BreakMeshByBlockGeneratorBase : public MeshGenerator
   /// this method save the boundary name/id pair
-  void mapBoundaryIdAndBoundaryName(boundary_id_type & /*boundaryID*/,
+  void mapBoundaryIdAndBoundaryName(const boundary_id_type & /*boundaryID*/,
                                     const std::string & /*boundaryName*/);
```
```suggestion
                                    const std::string & boundary_name);
```

ChengHauYang replied:
Fixed

lindsayad reviewed (commented):
just went through the headers for the moment. I'm very excited to see this de-hacking. Much appreciated @ChengHauYang !

ChengHauYang commented on framework/include/meshgenerators/BreakMeshByBlockGenerator.h:
```diff
@@ -59,11 +59,35 @@ class BreakMeshByBlockGenerator : public BreakMeshByBlockGeneratorBase
+
+  /// @brief Maps a pair of subdomain ids to the corresponding boundary id.
+  std::unordered_map<std::pair<subdomain_id_type, subdomain_id_type>, boundary_id_type>
+      _subid_pairs_to_boundary_id;
```
We intentionally keep two separate maps because they are populated at different stages and serve distinct purposes.

During `generate()`, we first collect interface sides and store them as `(subdomain_id_i, subdomain_id_j) -> (elem, side)` pairs. This stage only identifies which element sides belong to each block interface and does not create boundary IDs.

Boundary IDs are assigned later in addInterface(), once all "disjoint interface sidesets" have been gathered and synchronized across processors. At that point, we construct the map `(subdomain_id_i, subdomain_id_j) -> boundary_id`.

I also renamed `_new_boundary_sides_map` to `_subid_pairs_to_sides` to make it consistent with `_subid_pairs_to_boundary_id `.

ChengHauYang committed:
rename members and APIs base on Alex's feedback

ChengHauYang commented:
Hi @lindsayad and @GiudGiud, thank you both for the feedback. Since my family is visiting the U.S. for the first time, I’ll be starting a vacation tomorrow. I may not be able to address any additional comments promptly for the rest of December. Thanks for your understanding!

ChengHauYang committed:
Replace mesh-level incomplete interface checks with kernel-driven neighbor validation

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3459568/), step Results summary on a5dc8c6 wanted to post the following:

## Framework test summary

Compared against cc8efb1 in job [civet.inl.gov/job/3456352](https://civet.inl.gov/job/3456352).

### Added tests

| Test                                                                       | Time (s)   | Memory (MB)   |
|----------------------------------------------------------------------------|------------|---------------|
| `constraints/equal_value_boundary_constraint.error_multiple_primary_nodes` | 1.52       | 116.24        |
| `constraints/equal_value_boundary_constraint.error_not_find_node_id`       | 1.49       | 104.06        |
| `constraints/equal_value_boundary_constraint.set_primary_and_coord_error`  | 1.32       | 111.15        |
| `constraints/equal_value_boundary_constraint.test_primary_node_coord`      | 1.05       | 103.63        |
## Modules test summary

Compared against cc8efb1 in job [civet.inl.gov/job/3456352](https://civet.inl.gov/job/3456352).

### Added tests

| Test                                                                             | Time (s)   | Memory (MB)   |
|----------------------------------------------------------------------------------|------------|---------------|
| `solid_mechanics/test:cohesive_zone_model.czm_framework/boundary_pair_not_found` | 1.56       | 123.97        |

lindsayad commented on framework/include/loops/NonlinearThread.h:
```diff
@@ -76,6 +76,11 @@ class NonlinearThread : public ThreadedElementLoop<ConstElemRange>
   virtual void computeOnInternalFace() = 0;
   ///@}
 
+  /// Validate that treating the current side as external is acceptable when no neighbor is found.
```
Never mind, the base class doxygen and code make it clear what the use is. I'd remove this doxygen as it'll override the base class doxygen which is better in my opinion

lindsayad commented on framework/src/constraints/EqualValueBoundaryConstraint.C:
```diff
@@ -88,9 +94,60 @@ EqualValueBoundaryConstraint::updateConstrainedNodes()
+                      bd_node->_node->id()) != _secondary_node_ids.end() ||
+            bd_node->_bnd_id == _mesh.getBoundaryID(_secondary_node_set_id))
+        {
+          local_primary_node_ids.insert(bd_node->_node->id());
```
It seems like this conditional is determining whether a boundary node corresponds to the secondary node set, not the primary node set

hugary1995 replied:
I completed rewrote this part of EqualValueBoundaryConstraint. Many old constructs are outdated, and the logic became unclear as we accept changes over the years.

lindsayad commented on framework/src/meshgenerators/BreakMeshByBlockGenerator.C:
```diff
@@ -39,12 +40,20 @@ BreakMeshByBlockGenerator::validParams()
+  params.addRelationshipManager("ElementSideNeighborLayers",
+                                Moose::RelationshipManagerType::GEOMETRIC |
+                                    Moose::RelationshipManagerType::ALGEBRAIC |
+                                    Moose::RelationshipManagerType::COUPLING,
```
```suggestion
                                Moose::RelationshipManagerType::GEOMETRIC,
```
This object only needs geometric ghosting; it doesn't partake in algebraic calculations

lindsayad commented on framework/src/meshgenerators/BreakMeshByBlockGenerator.C:
```diff
@@ -136,210 +146,256 @@ BreakMeshByBlockGenerator::generate()
+      continue;
+
+    // If the node is not connected to any blocks, skip it
+    if (nodeid_to_connected_blocks.find(current_node_id) == nodeid_to_connected_blocks.end())
```
```suggestion
    auto it = nodeid_to_connected_blocks.find(current_node_id);
    if (it == nodeid_to_connected_blocks.end())
```

lindsayad commented on framework/src/meshgenerators/BreakMeshByBlockGenerator.C:
```diff
@@ -136,210 +146,256 @@ BreakMeshByBlockGenerator::generate()
+          // Get the two block IDs from the set
+          auto it = sets_blocks_for_this_node.begin();
+          subdomain_id_type block1 = *it;
+          subdomain_id_type block2 = *std::next(it);
```
```suggestion
          auto set_blocks_for_this_node_it = sets_blocks_for_this_node.begin();
          subdomain_id_type block1 = *set_blocks_for_this_node_it;
          subdomain_id_type block2 = *std::next(set_blocks_for_this_node_it);
```

lindsayad commented on framework/src/meshgenerators/BreakMeshByBlockGenerator.C:
```diff
@@ -349,26 +405,44 @@ BreakMeshByBlockGenerator::addInterfaceBoundary(MeshBase & mesh)
 
   BoundaryName boundary_name;
 
+  std::set<boundary_id_type> ids = boundary_info.get_boundary_ids();
```
```suggestion
  const std::set<boundary_id_type> & ids = boundary_info.get_boundary_ids();
```
I think this is ok

lindsayad commented on framework/src/meshgenerators/BreakMeshByBlockGenerator.C:
```diff
@@ -377,43 +451,70 @@ BreakMeshByBlockGenerator::addInterfaceBoundary(MeshBase & mesh)
+  for (auto & entry : _subid_pairs_to_boundary_id)
+  {
+    const auto & sub_pair = entry.first; // (subA, subB)
+    const auto & boundary_id = entry.second;
```
```suggestion
  for (const auto & [sub_pair, boundary_id] : _subid_pairs_to_boundary_id)
  {
```

lindsayad commented on framework/src/meshgenerators/BreakMeshByBlockGenerator.C:
```diff
@@ -437,3 +538,87 @@ BreakMeshByBlockGenerator::findBlockPairs(subdomain_id_type block_one, subdomain
+                                                                             // connected_blocks,
+                                                                             // ghost_pid)
+  std::map<processor_id_type, std::vector<NodeConnectedBlocksTuple>> to_owner;
+  for (const auto & map_entry : nodeid_to_connected_blocks)
```
structured binding

lindsayad commented on framework/src/meshgenerators/BreakMeshByBlockGenerator.C:
```diff
@@ -437,3 +538,87 @@ BreakMeshByBlockGenerator::findBlockPairs(subdomain_id_type block_one, subdomain
+    const Node * node = mesh.node_ptr(node_id);
+    if (node && node->processor_id() == mesh_pid)
+    {
+      const auto & blocks_set = nodeid_to_connected_blocks.at(node_id);
```
```suggestion
      const auto & blocks_set = libmesh_map_find(nodeid_to_connected_blocks, node_id);
```
In general we should never be using `map::at` because if it fails it doesn't give users any context. And oftentimes if we're using `at` it means we may have missed an opportunity to create and re-use an iterator

lindsayad commented on modules/solid_mechanics/test/tests/cohesive_zone_model/break_test.i:
AFAICT this is still unused

ChengHauYang replied:
Removed! Thanks!

lindsayad commented on modules/solid_mechanics/test/tests/cohesive_zone_model/czm_2DJunction_polycrystal.i:
same with this input

lindsayad commented on test/tests/constraints/equal_value_boundary_constraint/break_mesh_with_evbc.i:
```diff
@@ -0,0 +1,78 @@
+    fixed_normal = true
+    new_boundary = 'y1 x1 y0 x0'
+  []
+  parallel_type = REPLICATED
```
is this needed?

lindsayad commented on test/tests/meshgenerators/stitch_boundary_mesh_generator/stitch_3d.i:
```diff
@@ -68,6 +68,7 @@
     clear_stitched_boundary_ids = false
     stitch_boundaries_pair = '101 103'
   []
+  parallel_type = REPLICATED
```
why?

lindsayad reviewed (changes requested):
great work!

hugary1995 commented on framework/include/interfacekernels/InterfaceKernelBase.h:
```diff
@@ -43,6 +43,12 @@ class InterfaceKernelBase : public NeighborResidualObject,
+  bool requireNeighbor() const { return _require_neighbor; }
+
+  /// Set whether this interface kernel requires neighbor information
+  void requireNeighbor(bool req) { _require_neighbor = req; }
```
I suggested this change without thinking too much about it... That's true, and we should always error out if an interface kernel executes without neighbor.

hugary1995 commented on framework/include/mesh/MooseMesh.h:
```diff
@@ -1923,6 +1931,17 @@ class MooseMesh : public MooseObject, public Restartable, public PerfGraphInterf
+  /// element faces across the interface (e.g., for cohesive zone modeling).
+  ///
+  /// If only one sideset of an expected "+" / "-" pair is provided, the interface
+  /// sideset pair is considered incomplete.
```
While this perfect, we don't need this flag anymore. This flag was introduced to allow CZMAction to check and error. But we are now generalizing the check and moving it into the threaded loops.

hugary1995 committed:
Addressing comments (including my own); clean up EqualValueBoundaryConstraint

hugary1995 commented on framework/include/meshgenerators/BreakMeshByBlockGeneratorBase.h:
```diff
@@ -37,12 +37,12 @@ class BreakMeshByBlockGeneratorBase : public MeshGenerator
+                        const subdomain_id_type & /*secondaryBlockID*/,
+                        std::string & /*boundaryName*/,
+                        const boundary_id_type & /*boundaryID*/,
+                        BoundaryInfo & /*boundary_info*/);
```
There are also many other const-correctness issues in this method. Not all of them are your fault. I've fixed the ones Alex picked up. As I said I'll find some time to rewrite this mess.

hugary1995 commented on framework/include/loops/NonlinearThread.h:
```diff
@@ -130,6 +135,8 @@ class NonlinearThread : public ThreadedElementLoop<ConstElemRange>
   NonlinearSystemBase & _nl;
   unsigned int _num_cached;
+  mutable std::unordered_map<BoundaryID, std::vector<std::string>>
+      _boundary_neighbor_required_kernels;
```
No, actually I deleted this member as it's no longer needed. We just need to know whether the warehouse has any active interface kernel. This cache implementation isn't correct anyways, because we never invalidate the cache even when objects become active/inactive on the fly.

hugary1995 commented on framework/src/loops/NonlinearThread.C:
```diff
@@ -413,6 +414,49 @@ NonlinearThread::printBoundaryExecutionInformation(const unsigned int bid) const
+    auto cache_it = _boundary_neighbor_required_kernels.find(bnd_id);
+    if (cache_it == _boundary_neighbor_required_kernels.end())
+    {
+      std::vector<std::string> kernels;
```
I don't think we need this cache anymore.

hugary1995 committed:
Remove unused args

ChengHauYang committed:
tests: update BreakMeshByBlockGenerator/EVBC inputs, drop legacy CZM cases, and revert gold

- Drop default split_interface/add_interface_on_two_sides from BreakMeshByBlockGenerator test inputs
- Revert break_mesh_2DJunction_auto test
- Update EVBC break-mesh test to cover the multiple_primary_nodes error
- Fix EVBC test requirement wording (error vs warning)
- Revert interfacekernel and CZM gold files; regold only when the associated input files change due to EVBC updates
  (most tests should not fail unless pedantic checks are enabled; validation relies on physical-location mapping)

ChengHauYang commented on test/tests/constraints/equal_value_boundary_constraint/tests:
```diff
@@ -17,4 +27,25 @@
+    input = 'equal_value_boundary_constraint_test.i'
+    expect_err = "Couldn't find a node ID for the specified primary_node_coord."
+    cli_args = "Constraints/y_top/primary_node_coord='0.37 1.0 0.0'"
+    requirement = "The system shall report a warning when it cannot find a node ID for the specified node coordinates."
```
Fixed. Thanks!

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3485858/), step Results summary on 603938a wanted to post the following:

## Framework test summary

Compared against 4ac34d9 in job [civet.inl.gov/job/3483148](https://civet.inl.gov/job/3483148).

### Added tests

| Test                                                                       | Time (s)   | Memory (MB)   |
|----------------------------------------------------------------------------|------------|---------------|
| `constraints/equal_value_boundary_constraint.error_not_find_node_id`       | 1.29       | 96.97         |
| `constraints/equal_value_boundary_constraint.set_primary_and_coord_error`  | 1.27       | 104.58        |
| `constraints/equal_value_boundary_constraint.error_multiple_primary_nodes` | 1.26       | 109.01        |
| `constraints/equal_value_boundary_constraint.test_primary_node_coord`      | 0.87       | 113.55        |
## Modules test summary

Compared against 4ac34d9 in job [civet.inl.gov/job/3483148](https://civet.inl.gov/job/3483148).

### Added tests

| Test                                                                             | Time (s)   | Memory (MB)   |
|----------------------------------------------------------------------------------|------------|---------------|
| `solid_mechanics/test:cohesive_zone_model.czm_framework/boundary_pair_not_found` | 1.55       | 124.36        |

ChengHauYang committed:
[minor] make boundary name consistent

ChengHauYang commented:
Hi @GiudGiud, the original regold commits are still in the history. I did not rewrite history; I just restored a subset of gold files by checking them out from an earlier commit and committing the result.

ChengHauYang committed:
Add incomplete interface pair detection and CZM compatibility checks.

1. MooseMesh: Introduced `_has_incomplete_interface_pairs` flag with getter/setter, initialized in the constructor to detect missing reverse boundary pairs.
2. BreakMeshByBlockGenerator: Set `add_interface_on_two_sides = true` and `split_interface = true` by default; detects incomplete boundary pairs and outputs a warning if the user attempts to run CZM with such configurations.
3. CohesiveZoneAction & Tests: Added runtime error for incomplete CZM interfaces; updated test inputs to use explicit block-pair boundaries (replacing the unsupported single `interface` setup) and added the `boundary_pair_not_found` failure test

ChengHauYang committed:
update unit tests for new framework

ChengHauYang committed:
passing both replicated and distributed tests

ChengHauYang committed:
Framework: Coordinate-based EqualValueBoundaryConstraint node selection and removal of FakeNeighborRM

This update improves consistency and simplifies neighbor handling:

1. EqualValueBoundaryConstraint
   - Added `primary_node_coord` for selecting the primary node by coordinate, ensuring consistent behavior on distributed meshes where node IDs may differ.
   - If both `primary` and `primary_node_coord` are set, a `mooseError` is raised.
   - Added tests for coordinate-based selection and error handling.

2. Removed FakeNeighborRM
   - Eliminated the obsolete `FakeNeighborRM` used for ghosting in `BreakMeshByBlockGenerator`. libMesh ghosting now handles this.

ChengHauYang committed:
Ensure unique global primary-node selection under distributed meshes

Primary-node detection previously relied on summing local matches across ranks,
which incorrectly counted shared boundary nodes multiple times. This patch
replaces rank-local counting with a global deduplication of matched node IDs:
each rank reports its unique local matches, all IDs are gathered, a global
unique set is formed, and the smallest ID is chosen to ensure deterministic
selection. This guarantees correct behavior regardless of mesh partitioning.

ChengHauYang committed:
(a) address comments (b) fix bugs related to disjointed pair not found

ChengHauYang committed:
Remove unused variable

ChengHauYang committed:
fix one test: 1d_interface.no_failed_point_inversions

ChengHauYang committed:
regold files associated with EVBC

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/3488220/), step Verify coverage on 7652bc6 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `framework` coverage rate (required: 81.0%)
- Failed to generate `chemical_reactions` coverage rate (required: 92.0%)

ChengHauYang commented:
Hi @GiudGiud, I’ve finished rewriting the history to remove the redundant Exodus files. Thanks!

lindsayad committed:
Protect against locally empty _connected_nodes when picking primary node id

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3491849/), step Results summary on 231229c wanted to post the following:

## Framework test summary

Compared against 4545b09 in job [civet.inl.gov/job/3489297](https://civet.inl.gov/job/3489297).

### Added tests

| Test                                                                       | Time (s)   | Memory (MB)   |
|----------------------------------------------------------------------------|------------|---------------|
| `constraints/equal_value_boundary_constraint.error_multiple_primary_nodes` | 1.50       | 117.51        |
| `constraints/equal_value_boundary_constraint.set_primary_and_coord_error`  | 1.41       | 99.19         |
| `constraints/equal_value_boundary_constraint.error_not_find_node_id`       | 1.32       | 103.88        |
| `constraints/equal_value_boundary_constraint.test_primary_node_coord`      | 0.95       | 121.16        |
## Modules test summary

Compared against 4545b09 in job [civet.inl.gov/job/3489297](https://civet.inl.gov/job/3489297).

### Added tests

| Test                                                                             | Time (s)   | Memory (MB)   |
|----------------------------------------------------------------------------------|------------|---------------|
| `solid_mechanics/test:cohesive_zone_model.czm_framework/boundary_pair_not_found` | 1.57       | 122.06        |

hugary1995 commented:
See #32192 for a patch.
